### ISSUE TITLE
[FR] Change order of "preserve chrominance" norms in filmic rgb 

### ISSUE THREAD

**Issue by EC1000** [darktable-org/darktable#12365]
**Is your feature request related to a problem? Please describe.**
In filmic rgb options tab, norms for "preserve chrominance" are currently listed in this order: 
> no
> max RGB
> luminance Y
> RGB power norm
> RGB euclidean norm
> RGB euclidean norm (legacy)

When I change the norm to compare the different results they give, if I do it by scrolling with the mouse wheel (or a 'next norm' shortcut), the impact on the picture may go from one extreme to another (e.g. 'no' vs 'max rgb'). 

Of course it is always possible to clic on the list of norms and choose one by clicking on it! :) I just find it is a little longer and less convenient (and not possible with a knob on a MIDI controller).

**Describe the solution you'd like**
I would I would like to suggest changing this GUI order, so that scrolling throught it has a more gradual impact on the picture: 
- more chroma to less chroma in highlights 
- less constrast to more contrast

So something like: 
> max RGB
> RGB power norm
> RGB euclidean norm
> luminance Y
> no
> RGB euclidean norm (legacy)

I know my description simplifies the impact on the picture (it is not just chroma in highlights and contrast) and it also depends on the colors on the pictures (for example on reds, 'luminance Y' is probably more contrasty than 'no'). But nevertheless, I find this re-order should work well in most cases. 

**Alternatives**
Define several shortcuts to jumb to a specific norm, and order these shortcuts in the "logical way" I want.

**Comment by EC1000** [comment]
As a side note, I don't know if the old "RGB euclidean norm (legacy)" could be remove **from the gui** (like a depreciated module); but if it possible, I would find it better to remove it (see PR that introduces the new euclidiean norm  https://github.com/darktable-org/darktable/pull/8272 about the original flaw of this norm).

**Comment by Jiyone** [comment]
> I don't know if the old "RGB euclidean norm (legacy)" could be remove **from the gui** (like a depreciated module)

I believe this will not be possible, the module needs to be backward compatible with old edit.

I don’t think this will be a serious benefit to sort norms in visual effect order…
One should know the effect just by reading the name 👀

**Comment by victoryforce** [comment]
> > I don't know if the old "RGB euclidean norm (legacy)" could be remove **from the gui** (like a depreciated module)
> 
> I believe this will not be possible, the module needs to be backward compatible with old edit.

Removing from the GUI is not the same as removing the implementation code that supports the old edits. So it is possible. Whether it is better to do so is another question.

**Comment by Nilvus** [comment]
I agree with @Jiyone: I see no benefit on this, result depends a lot of the image also. For example, difference is often huge between some norms and luminance one when the image contains a lot of a blue sky, but on some images, difference is more subtle.

Thus, I'm quite sure @aurelienpierre put this order on purpose.

**Comment by EC1000** [comment]
> One should know the effect just by reading the name :eyes:

I usually know what aspect it will increase or decrease -- for example, local constrast on skins or chromaticity on skies -- but I don't know how much and what will be my preferred result before seeing it and comparing the possibilities. 

I also usually know in which order I have to test them to see a gradual effect, but to test the 5 possibilites **with this gradual order**, with a mouse you will need to: (1) clic on the menu, (2) moove you mouse, (3) clic on the chosen norm, (4) go back to the menu to iterate. If the order were changed, you could just: (1) use your mouse wheel on the menu. Or: (1) use a shortcut (a 'next' button, or an encoder on a MIDI device, that's what I use). 


I agree with you, and already mentionned it in my first post, that my so-called "gradual order" may not always hold, because the effect may differ depending on colors. 
But at least this order would be gradual for two major usecases, that are skies and skins: 
- chromaticity on light blues (so skies) will be better retains with max RGB > RGB power norm > RGB euclidean norm > luminance Y > No 
- local constrat on reds (in particular skins) will increase with this order


To quote the usermanual https://docs.darktable.org/usermanual/4.0/en/module-reference/processing-modules/filmic-rgb/: 
> **preserve chrominance**
> - max RGB (...) tends to darken the blues, especially skies (...). It can also flatten the local contrast somewhat.
> - luminance Y (...) tends to darken and increase local contrast in the reds
> - RGB power norm (...) is usally a good compromise between the max RGB and the luminance Y values.
> - RGB euclidean norm (...) weighs more heavily on highlights than the power norm and gives more highlights desaturation

So the gradation between the different norms is already exposes in the usermanual. So my proposition is basically to translate this in the GUI.

**Comment by EC1000** [comment]
> Thus, I'm quite sure @aurelienpierre put this order on purpose.

Regarding the reason of the current order, only AP could tell, but I seems to me it is mainly 'historical' reasons (i.e. the order they were added):  
- From an article on the initial filmic module https://eng.aurelienpierre.com/2018/11/filmic-darktable-and-the-quest-of-the-hdr-tone-mapping/ (this advice is not 'valid' anymore I would say :smile: but it probably explains why there is 'No' as a first choice): 
> An efficient workflow would be to start setting the module with chroma preservation disabled, then, as a last step, enable it to see if it brings an additional benefit to your picture.


- still from the usermanual: 
> max RGB (...) is the same behaviour as the original version of the filmic rgb module.

(note: I think it should read "original version of the filmic module" - without "rgb" - because in the original filmic module there was just a "preserve the chrominance" tick, without several norms)

**Comment by SoupyGit** [comment]
I strongly agree with this proposal. For sunset images (or any image with colourful highlights) I always want to compare max RGB with RGB power norm. It is MUCH more convenient if they are right next to each other on the list (when toggling through using up and down keys - as you do if you have set a shortcut for this feature). Otherwise, one has to click from one to the other using mouse, which is not the end of the world, but not as fast for power users. If one wants to click from one to the other using a shortcut, they have to go shortcut > click > shortcut > click. An inconvenience. Otherwise we have to scroll max RGB > luminance y > RGB power norm, where luminance y always prevents an easy comparison between max RGB and RGB power norm.

**Comment by EC1000** [comment]
A change in the order of methods and hiding an obsolete method was done for the highlight reconstruction by @jenshannoschwalm in https://github.com/darktable-org/darktable/pull/12623. 

What I suggest about removing **legacy** RGB euclidean norm from GUI seems to follow a similar logic. Maybe @flannelhead you could comment on this? reading what you wrote in #8272, it seems flawed (and have never seen a picture where this legacy norm was better).

**Comment by flannelhead** [comment]
> removing legacy RGB euclidean norm from GUI

To me, this seems like a good idea. For anyone who would like to implement this, it would probably make sense to do it like @jenshannoschwalm did for highlight reconstruction in https://github.com/darktable-org/darktable/pull/12623.

**Comment by Nilvus** [comment]
@flannelhead: I'm wondering if this should not be part of PR related to your filmic proposal on #12847? Would be simpler. Just note that feature freeze for 4.2 is in 2 days.

**Comment by Nilvus** [comment]
Regarding comments here and discussion on #12847 + highlight reconstruction work done, if order is changed, I think this should be:

- RGB power norm
- max RGB
- RGB euclidean norm
- luminance Y
- no
- RGB euclidean norm (legacy)

As RGB power norm seems to be soon default setting, better to have it first, then EC1000 proposal on first comment here. And so last line hidden if not used on the image.

**Comment by flannelhead** [comment]
@Nilvus I'll see if I can make it to the freeze. However, I tend to disagree about the order. It has not previously been the case that the default is first, and would kind of defeat the point of "ordering by effect" that is proposed here.

**Comment by Nilvus** [comment]
No problem. I was first against changing order. But all comments changed my mind. Said that, actual order is great to me. No need to changed it. If it's changed, no problem also to me.

**Comment by SoupyGit** [comment]
Only difference between @Nilvus order and @EC1000 order is swapping placement of Max rgb and power norm. If we agree default does not need to be first, and that power norm effect is between Max rgb and euclidean norm, then we now have three of us agreeing about the proposed changed. Is there anyone still standing in opposition?

**Comment by Nilvus** [comment]
I do not say I agree. I say "if order is changed". I'm just no more against change but I still think changing order is not needed.

**Comment by aurelienpierre** [comment]
This is bikeshedding.

**Comment by EC1000** [comment]
@flannelhead do you think the order could be change now? I read what you wrote in https://github.com/darktable-org/darktable/pull/12882 about the dev cycle of previous release/next release, I suppose the timing would be ok for this cycle...? 
It is probably bike-shedding, but so are many other small improvements! :wink:

**Comment by flannelhead** [comment]
I'm not currently interested in doing this change. If this change is desired, this should be quite an easy task for someone willing to start contributing.

**Comment by dterrahe** [comment]
> I always want to compare max RGB with RGB power norm. It is MUCH more convenient if they are right next to each other on the list (when toggling through using up and down keys - as you do if you have set a shortcut for this feature). Otherwise, one has to click from one to the other using mouse, which is not the end of the world, but not as fast for power users. If one wants to click from one to the other using a shortcut, they have to go shortcut > click > shortcut > click. An inconvenience. Otherwise we have to scroll max RGB > luminance y > RGB power norm, where luminance y always prevents an easy comparison between max RGB and RGB power norm.

This is indeed bikeshedding, in the sense that people have different, irreconcilable, preferences. Which is why I tried to make the shortcut system configurable enough that everybody can get what they want. In this case, if you define a scroll shortcut with _speed_ 2, you can easily skip over the intermediate option. But you'd have to be careful to not scroll too far, in order to avoid the no-luminance-euclidean sequence rather than the max-power sequence. So what I'd do is enable fallbacks, define a "normal" (single key press) shortcut to preserve chrominance (which would trigger a popup, so you can directly select max RGB) and then, while holding that shortcut, scroll down/up/down/up to compare the two.

As an aside, you can also customize per-widget speeds. If you do this for a slider, it also affects the speed with which the arrow keys or the scroll wheel over the slider changes the value, not just shortcuts. This hasn't been implemented for combos because, well, nobody cares. Haven't even checked but it should be the trivialest of changes.

**Comment by ralfbrown** [comment]
Closing because "preserve chrominance" norms are no longer an option in filmic v7, which has been the default for multiple releases now, and there are other ways of achieving the desired result for those still using older color science versions.
